source/link/linker.cpp

```diff
@@ -16,1183 +16,277 @@
+  for (const auto& i : exports)
+    std::cout << "Export: " << i.second << ", " << i.first << "\n";
+
+  std::unordered_map<spv::Id, spv::Id> linking_table;
+  linking_table.reserve(imports.size());
+  for (const auto& i : imports) {
+    auto j = exports.find(i.second);
+    if (j == exports.end()) {
+      std::cerr << "Failed to find \"" << i.second << "\"\n";
+      return 1;
     }
-    if (j == length)
-      prev_j = j;
-
-    std::memcpy(merged_binary + merged_length, binary + index,
-                (prev_j - index) * sizeof(spirv_word));
-    merged_length += prev_j - index;
-    indices[i] = j;
+    linking_table.emplace(i.first, j->second);
   }
 
-  // It is up to the driver to solve builtins that are imported.
-  for (const auto k : builtins)
-    imports.erase(k);
+  for (const auto& i : linking_table)
+    std::cout << i.first << " -> " << i.second << "\n";
 
-  for (const auto &k : imports) {
-    const auto exports_iter = std::find_if(exports.begin(), exports.end(), [&k](const std::pair<spv::Id, const char *> &p){
-      return !strcmp(k.second, p.second);
-    });
-    if (exports_iter == exports.end()) {
-      return libspirv::DiagnosticStream(position, context.consumer,
-                                     SPV_ERROR_INVALID_DATA)
-             << "Missing symbol \"" << k.second << "\"";
+  // Remove prototypes of imported functions
+  for (auto i = linkedModule->begin(); i != linkedModule->end();) {
+    const auto function_id = i->DefInst().result_id();
+    if (imports.find(function_id) == imports.end()) {
+      ++i;
+      continue;
     }
-    link_table.emplace(k.first, exports_iter->first);
+    i = i.Erase();
   }
 
-  // Copy types/constants/global variables/OpUndef
-  for (uint32_t i = 0u; i < num_binaries; ++i) {
-    const uint32_t* const binary = binaries[i].data();
-    const size_t length = binaries[i].size();
-    uint32_t index = indices[i];
-    uint32_t j = index;
-    uint32_t prev_j = j;
-    while (j < length) {
-      uint32_t word_count;
-      const spv::Op opcode = ParseOpcode(binary, j, word_count);
-
-      prev_j = j;
-
-      spv::StorageClass storage_type = static_cast<spv::StorageClass>(GetWord(binary, j + 3u));
-      if (!IsTypeOrConstantOpcode(opcode) ||
-          (opcode == spv::Op::OpVariable &&
-           storage_type == spv::StorageClass::Function))
-        break;
-
-      j += word_count;
-
-      // Remove imported variables
-      spv::Id tmp = static_cast<spv::Id>(GetWord(binary, prev_j + 2u));
-      if (opcode == spv::Op::OpVariable &&
-         link_table.find(tmp) != link_table.end()) {
-        std::memcpy(merged_binary + merged_length, binary + index,
-                    (prev_j - index) * sizeof(spirv_word));
-        merged_length += prev_j - index;
-        index = j;
-      }
+  // Remove declarations of imported global variables
+  for (auto i = linkedModule->types_values_begin(); i != linkedModule->types_values_end();) {
+    if (i->NumOperands() < 2u) {
```

Comment:
This is a bit mysterious.    Better to just compare opcode to OpVariable.

Suggest
```
if (i->opcode() == OpVariable) { .... do stuff ... }
```

I tend to prefer avoiding non-local control flow like continue and break when the straightforward expression is slightly more compact.  Framing in the positive is nicer, provided indentation remains sensibly bounded.